prog/dftb+/lib_dftb/populations.F90

```diff
@@ -451,5 +452,86 @@ subroutine getChargePerShell(qOrb, orb, species, chargePerShell, qRef)
   end subroutine getChargePerShell
 
 
+  !> Block mulliken analysis with dense lower triangle matrices.
+  subroutine denseBlockMulliken(rhoSqr, overSqr, q0, iSquare, qq)
+
+    !> Square (lower triangular) spin polarized density matrix
+    real(dp), intent(in) :: rhoSqr(:,:,:)
+
+    !> Square (lower triangular) overlap matrix
+    real(dp), intent(in) :: overSqr(:,:)
+
+    !> reference charges
+    real(dp), intent(in) :: q0(:,:,:)
+
+    !> Atom positions in the row/column of square matrices
+    integer, intent(in) :: iSquare(:)
+
+    !> Mulliken block charges on output (mOrb, mOrb, nAtom, nSpin)
+    real(dp), intent(out) :: qq(:,:,:,:)
+
+    real(dp), allocatable :: tmpS(:,:)
+    real(dp), allocatable :: tmpD(:,:,:)
+    real(dp), allocatable :: tmp_mat(:,:)
+    real(dp), allocatable :: tmp_block(:,:)
+
+    integer :: nAOs, nat, nSpin, ii, jj, kk, iat, iS
+
+    nSpin = size(rhoSqr, dim=3)
+    nat = size(iSquare, dim=1) - 1
+    nAOs = size(rhoSqr, dim=1)
+
+    allocate(tmpS(nAOs,nAOs))
+    allocate(tmpD(nAOs,nAOs,nSpin))
+    allocate(tmp_mat(nAOs,nAOs))
+    allocate(tmp_block(nAOs,nAOs))
+
+    ! ... symmetrize overlap and density matrices
+    tmpS(:,:) = 0.0_dp
+    tmpS(:,:) = overSqr + transpose(overSqr)
+    do ii = 1, nAOs
+      tmpS(ii,ii) = tmpS(ii,ii) - overSqr(ii,ii)
+    end do
+
+    ! ... tmpD has 1:alpha, 2:beta
+    do iS = 1, nSpin
+      tmpD(:,:,iS) = 0.0_dp
+      tmpD(:,:,iS) = rhoSqr(:,:,iS) + transpose(rhoSqr(:,:,iS))
+      do ii = 1, nAOs
+        tmpD(ii,ii,iS) = tmpD(ii,ii,iS) - rhoSqr(ii,ii,iS)
+      end do
+    end do
+
+    qq(:,:,:,:) = 0.0_dp
+    do iS = 1, nSpin
+
+      tmp_block(:,:) = 0.0_dp
+      tmp_mat(:,:) = 0.0_dp
+      call gemm(tmp_mat,tmpS,tmpD(:,:,iS),alpha=1.0_dp,beta=0.0_dp,transA='N',transB='N')
+      tmp_block(:,:) = tmp_block(:,:) + 0.5_dp * tmp_mat(:,:)
+
+      tmp_mat(:,:) = 0.0_dp
+      call gemm(tmp_mat,tmpD(:,:,iS),tmpS,alpha=1.0_dp,beta=0.0_dp,transA='N',transB='N')
+      tmp_block(:,:) = tmp_block(:,:) + 0.5_dp * tmp_mat(:,:)
+
+      do iat = 1, nat
```

Comment:
Should this be done in the population evaluation routine? This is the gross electron count not the charge on the orbitals.